lighthouse-core/audits/robust-csp.js

```diff
@@ -0,0 +1,188 @@
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      id: 'robust-csp',
+      title: str_(UIStrings.title),
+      failureTitle: str_(UIStrings.failureTitle),
+      description: str_(UIStrings.description),
+      requiredArtifacts: ['devtoolsLogs', 'MetaElements', 'URL'],
+    };
+  }
+
+  /**
+   * @param {Finding} finding
+   * @return {LH.Audit.Details.TableItem}
+   */
+  static findingToTableItem(finding) {
+    return {
+      directive: finding.directive,
+      description: getTranslatedDescription(finding),
+    };
+  }
+
+  /**
+   * @param {Array<string>} rawCsps
+   * @return {LH.Audit.Details.TableItem[]}
+   */
+  static collectSyntaxResults(rawCsps) {
+    /** @type {LH.Audit.Details.TableItem[]} */
+    const results = [];
+
+    const syntaxFindingsByCsp = evaluateRawCspForSyntax(rawCsps);
+    for (let i = 0; i < rawCsps.length; ++i) {
+      const items = syntaxFindingsByCsp[i].map(this.findingToTableItem);
+      if (!items.length) continue;
+
+      const description = str_(UIStrings.syntaxMessage, {
+        numSyntax: items.length,
+        rawCsp: rawCsps[i],
+      });
+
+      results.push({
+        description,
+        subItems: {
+          type: 'subitems',
+          items,
+        },
+      });
+    }
+
+    return results;
+  }
+
+  /**
+   * @param {Array<string>} cspHeaders
+   * @param {Array<string>} cspMetaTags
+   * @return {LH.Audit.Details.TableItem[]}
+   */
+  static collectVulnerabilityResults(cspHeaders, cspMetaTags) {
+    const rawCsps = [...cspHeaders, ...cspMetaTags];
+    const findings = evaluateRawCspForFailures(rawCsps);
+    return findings.map(this.findingToTableItem);
+  }
+
+  /**
+   * @param {Array<string>} cspHeaders
+   * @param {Array<string>} cspMetaTags
+   * @return {LH.Audit.Details.TableItem[]}
+   */
+  static collectSuggestionResults(cspHeaders, cspMetaTags) {
+    const rawCsps = [...cspHeaders, ...cspMetaTags];
+    const findings = evaluateRawCspForWarnings(rawCsps);
+    const results = [
+      ...findings.map(this.findingToTableItem),
+      ...this.collectSyntaxResults(rawCsps),
+    ];
+    if (cspMetaTags.length) {
+      results.push({description: str_(UIStrings.metaTagMessage)});
+    }
+    return results;
+  }
+
+  /**
+   * @param {LH.Artifacts} artifacts
+   * @param {LH.Audit.Context} context
+   * @return {Promise<LH.Audit.Product>}
+   */
+  static async audit(artifacts, context) {
+    const devtoolsLog = artifacts.devtoolsLogs[Audit.DEFAULT_PASS];
+    const mainResource = await MainResource.request({devtoolsLog, URL: artifacts.URL}, context);
+
+    const cspMetaTags = artifacts.MetaElements
+      .filter(m => {
+        return m.httpEquiv && m.httpEquiv.toLowerCase() === 'content-security-policy';
+      })
+      .map(m => m.content || '');
+    const cspHeaders = mainResource.responseHeaders
+      .filter(h => {
+        return h.name.toLowerCase() === 'content-security-policy';
+      })
+      .map(h => h.value);
+
```

Comment:
Should empty CSPs be filtered out, so we say "there is no CSP" if it's just an empty string; instead of saying various things about how `''` is not robust.

Reply:
Yeah, that makes sense to me.